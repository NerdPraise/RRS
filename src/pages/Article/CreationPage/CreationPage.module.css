:local(.CreationPage) {
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
}

:local(.container) {
  min-width: 350px;
  max-width: 700px;
}

:local(.CreationPage .name__group) {
  display: flex;
  justify-content: space-between;
}

:local(.tag) {
  text-align: center;
  font-weight: bold;
  margin-bottom: 10px;
}

:local(.header) {
  text-align: center;
}

:local(.input__cont) {
  margin-bottom: 3px;
}
