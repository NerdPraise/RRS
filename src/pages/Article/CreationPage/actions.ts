// No need for API definitions

export const uploadArticle = (data: FormData) => {
  return fetch('https://jsonplaceholder.typicode.com/posts/', {
    method: 'POST',
    body: data,
  })
}
