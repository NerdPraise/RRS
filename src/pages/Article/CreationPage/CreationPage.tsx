import { FormEvent, useRef, useState } from 'react'
import { Input, Toast } from '@Components'

import { uploadArticle } from './actions'

import style from './CreationPage.module.css'

export const CreationPage = () => {
  const [formError, setFormError] = useState<{ [key: string]: string }>({})
  const [statusMessage, setStatusMessage] = useState<{
    message?: string
    level?: string
  }>()
  const formRef = useRef(null)

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    setStatusMessage({})
    e.preventDefault()
    const form = e.target as HTMLFormElement
    const data = new FormData(form)

    uploadArticle(data)
      .then(res => res.json())
      .then(_ => {
        form.reset()
        setStatusMessage({ message: 'Successful request', level: 'success' })
      })
      .catch(err => setStatusMessage({ message: err.message, level: 'error' }))
  }

  const shouldDisable = !!Object.keys(formError).length

  return (
    <div className={style.CreationPage}>
      <Toast
        message={statusMessage?.message}
        level={statusMessage?.level as 'error' | 'success'}
      />
      <div className={style.container}>
        <h2 className={style.header}>Article Page</h2>
        <div className={style.tag}>Create your new articles</div>
        <div className="form">
          <form ref={formRef} onSubmit={handleSubmit} role="form">
            <div className={style.name__group}>
              <div className="input__group half">
                <label htmlFor="first_name ">First name</label>
                <Input
                  type="text"
                  name="first_name"
                  error={formError}
                  setError={setFormError}
                  classname={style.input__cont}
                />
              </div>
              <div className="input__group half">
                <label htmlFor="last_name ">Last name</label>
                <Input
                  type="text"
                  name="last_name"
                  error={formError}
                  setError={setFormError}
                  classname={style.input__cont}
                />
              </div>
            </div>
            <div className="input__group">
              <label htmlFor="phone">Phone</label>
              <Input
                type="tel"
                name="phone"
                error={formError}
                setError={setFormError}
                classname={style.input__cont}
              />
            </div>
            <div className="input__group">
              <label htmlFor="email">Email</label>
              <Input
                type="email"
                name="email"
                error={formError}
                setError={setFormError}
                classname={style.input__cont}
              />
            </div>
            <div className="input__group">
              <label htmlFor="address">Address</label>
              <Input
                type="text"
                name="address"
                error={formError}
                setError={setFormError}
                classname={style.input__cont}
              />
            </div>
            <div className="input__group">
              <label htmlFor="body">Body</label>
              <textarea
                name="body"
                rows={12}
                cols={60}
                required
                role="textbox"
              />
            </div>
            <button disabled={shouldDisable} data-testid="textarea">
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  )
}
