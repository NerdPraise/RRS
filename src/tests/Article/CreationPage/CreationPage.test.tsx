import { describe, expect, it, vi, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import '@testing-library/jest-dom/vitest'
import React from 'react'

import { CreationPage } from '@Pages/Article/CreationPage'

React.createElement('div', { id: 'portal' })

vi.mock('react-dom', () => ({
  ...vi.importActual('react-dom'),
  createPortal: (children: JSX.Element) => children,
}))

describe('CreationPage Renders', () => {
  afterEach(() => {
    cleanup()
  })

  it('renders correctly', () => {
    const page = render(<CreationPage />)
    expect(page.queryByText('First name')).toBeInTheDocument()
    expect(page.queryByText('Body')).toBeInTheDocument()
    expect(page.getByRole('button')).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const page = render(<CreationPage />)
    expect(page).toMatchSnapshot()
  })
})
