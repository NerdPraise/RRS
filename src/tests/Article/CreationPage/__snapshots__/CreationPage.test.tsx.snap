// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreationPage Renders > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_CreationPage_3c6616"
      >
        <div
          class="_container_3c6616"
        >
          <h2
            class="_header_3c6616"
          >
            Article Page
          </h2>
          <div
            class="_tag_3c6616"
          >
            Create your new articles
          </div>
          <div
            class="form"
          >
            <form
              role="form"
            >
              <div
                class="_name__group_3c6616"
              >
                <div
                  class="input__group half"
                >
                  <label
                    for="first_name "
                  >
                    First name
                  </label>
                  <div
                    class="_input__cont_3c6616 _Input_98f327"
                  >
                    <input
                      data-testid="first_name"
                      name="first_name"
                      type="text"
                    />
                  </div>
                </div>
                <div
                  class="input__group half"
                >
                  <label
                    for="last_name "
                  >
                    Last name
                  </label>
                  <div
                    class="_input__cont_3c6616 _Input_98f327"
                  >
                    <input
                      data-testid="last_name"
                      name="last_name"
                      type="text"
                    />
                  </div>
                </div>
              </div>
              <div
                class="input__group"
              >
                <label
                  for="phone"
                >
                  Phone
                </label>
                <div
                  class="_input__cont_3c6616 _Input_98f327"
                >
                  <input
                    data-testid="phone"
                    name="phone"
                    type="tel"
                  />
                </div>
              </div>
              <div
                class="input__group"
              >
                <label
                  for="email"
                >
                  Email
                </label>
                <div
                  class="_input__cont_3c6616 _Input_98f327"
                >
                  <input
                    data-testid="email"
                    name="email"
                    type="email"
                  />
                </div>
              </div>
              <div
                class="input__group"
              >
                <label
                  for="address"
                >
                  Address
                </label>
                <div
                  class="_input__cont_3c6616 _Input_98f327"
                >
                  <input
                    data-testid="address"
                    name="address"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="input__group"
              >
                <label
                  for="body"
                >
                  Body
                </label>
                <textarea
                  cols="60"
                  name="body"
                  required=""
                  role="textbox"
                  rows="12"
                />
              </div>
              <button
                data-testid="textarea"
              >
                Submit
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_CreationPage_3c6616"
    >
      <div
        class="_container_3c6616"
      >
        <h2
          class="_header_3c6616"
        >
          Article Page
        </h2>
        <div
          class="_tag_3c6616"
        >
          Create your new articles
        </div>
        <div
          class="form"
        >
          <form
            role="form"
          >
            <div
              class="_name__group_3c6616"
            >
              <div
                class="input__group half"
              >
                <label
                  for="first_name "
                >
                  First name
                </label>
                <div
                  class="_input__cont_3c6616 _Input_98f327"
                >
                  <input
                    data-testid="first_name"
                    name="first_name"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="input__group half"
              >
                <label
                  for="last_name "
                >
                  Last name
                </label>
                <div
                  class="_input__cont_3c6616 _Input_98f327"
                >
                  <input
                    data-testid="last_name"
                    name="last_name"
                    type="text"
                  />
                </div>
              </div>
            </div>
            <div
              class="input__group"
            >
              <label
                for="phone"
              >
                Phone
              </label>
              <div
                class="_input__cont_3c6616 _Input_98f327"
              >
                <input
                  data-testid="phone"
                  name="phone"
                  type="tel"
                />
              </div>
            </div>
            <div
              class="input__group"
            >
              <label
                for="email"
              >
                Email
              </label>
              <div
                class="_input__cont_3c6616 _Input_98f327"
              >
                <input
                  data-testid="email"
                  name="email"
                  type="email"
                />
              </div>
            </div>
            <div
              class="input__group"
            >
              <label
                for="address"
              >
                Address
              </label>
              <div
                class="_input__cont_3c6616 _Input_98f327"
              >
                <input
                  data-testid="address"
                  name="address"
                  type="text"
                />
              </div>
            </div>
            <div
              class="input__group"
            >
              <label
                for="body"
              >
                Body
              </label>
              <textarea
                cols="60"
                name="body"
                required=""
                role="textbox"
                rows="12"
              />
            </div>
            <button
              data-testid="textarea"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
